lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1673,6 +1674,42 @@ bool TypeChecker::typeCheckFunctionBodyUntil(FuncDecl *FD,
   BraceStmt *BS = FD->getBody();
   assert(BS && "Should have a body");
 
+  // FIXME: FIXME_NOW: DETERMINE: What is the appropriate place for this code to
+  //        live?
+  if (FD->hasSingleExpressionBody()) {
+    auto resultTypeLoc = FD->getBodyResultTypeLoc();
+    auto E = FD->getSingleExpressionBody();
+
+    // FIXME: FIXME_NOW: DETERMINE: Does this actually need to be here?  It's 
+    //        being done for closures, but there may be a reason that applies
+    //        there but not here.
+    //
+    //           if (E != FD->getSingleExpressionBody()) {
+    //             FD->setSingleExpressionBody(E);
+    //           }
+
+    if (resultTypeLoc.isNull() || resultTypeLoc.getType()->isVoid()) {
+      BS->setElement(0, E);
+    } else {
+      // FIXME: FIXME_NOW: DETERMINE: Is this the appropriate mechanism to use
+      //        to divine the type of E?
+      preCheckExpression(E, FD);
+      constraints::ConstraintSystem cs(*this, FD, 
+                                       constraints::ConstraintSystemFlags::SuppressDiagnostics);
+      SmallVector<constraints::Solution, 4> viable;
+      if(!cs.solve(E, /*convertType*/Type(), nullptr, viable,
+                   FreeTypeVariableBinding::Disallow)) {
+        auto &solution = viable[0];
```

Comment:
This still weirds me out. How do we know there's a solution?

Reply:
If I understand correctly, solve will return true if we don't get any solutions:

```
/// \returns true is an error occurred, false is system is consistent
/// and solutions were found.
```

Reply:
\>< That's what I get for not reading the docs. I think there's still a *small* hole here where we have overloads that may or may not be the same as the result type, which means multiple solutions. In that case the contextual type will make a difference.

Reply:
Is this the kind of scenario you were thinking of, where overloads with different return types would mean multiple solutions?

```
enum MyOwnNever {}
func die() -> MyOwnNever { print("MyOwnNever"); fatalError() }
func die() -> Never { print("Never"); fatalError() }
func die() -> Int { 6 }

func roll() -> Int { die() }
func roll2() -> Never { die() }
func roll3() -> MyOwnNever { die() }
```

It turns out that it works (and that `viable.size()` is always 1).  Could it be that the providing `FD` as the `DeclContext` to the `ConstraintSystem` is providing the context necessary to disambiguate?

Reply:
> Is this the kind of scenario you were thinking of, where overloads with different return types would mean multiple solutions?

Not quite. There's a scenario where one choice *is* better (because of other overload resolution rules), but it's not the "right" one.

```
func evil() -> Never { print("Never"); fatalError() }
func evil(extraArgument: Int = 0) -> Int { return extraArgument }

func roll() -> Int { evil() }
```

However…I'm realizing that interpreting this as `return evil()` would be a source-breaking change, which means we have to make a choice as to whether that's allowed. So one way to go here is a funny extra rule in single-expression functions that's different from single-expression closures (possibly conditional on language mode): "unlike with an explicit `return`, overload resolution does not take the calling function's return type into account; it's only implied if the types match up".

(I have to eat crow now because in a previous discussion with other Apple folks I said "there exists an implementation that's not source-breaking at all", and there does, but it's unfortunate that that implementation isn't going to follow the same rules as we get for closures.)

> Could it be that the providing `FD` as the `DeclContext` to the `ConstraintSystem` is providing the context necessary to disambiguate?

I don't think so. I guess I'll have to bother a type-checker person, or one of us has to track down why it's working.

Reply:
Does the source compatibility breakage only happen if the preferred overload returns `Never`?  If so, could the difference between implicit returns for functions and closures be specific to overload resolution when one of the candidates returns `Never`?

It would be nice to keep support for

```
func zero() -> String { "" }
func zero(plus other: Int = 0) -> Int { other }
func identity() -> Int { zero() }
```

where none of the candidates return Never since selecting the one preferred by overload resolution rules wouldn't have been legal code without implicit return.

That would make the rule even more complicated, but maybe it would be worth it?  Overloading where one of the candidates returns `Never` might be less common (than overloading where none of the candidates returns `Never`) enough that it would be worth the rule for that case being different.

Reply:
`Never` or something `Never`-ish, yeah. But that seems harder to implement if you go the direction I suggest where we manage to only type-check the expression one way.

Reply:
If we stick with the two-pass approach (with writing back the prechecked expression back into the single expression body of the function), it seems like we get the best of both worlds.  

I looked at a few permutations of your example:

```
func evil_1(value: Int = 0) -> Int { return 0 }
func evil_1() -> Never { fatalError() }

func pass_1_noreturn() -> Int { evil_1() }
func pass_1_return() -> Int { return evil_1() }

func evil_2() -> Int { return 0 }
func evil_2(value: Int = 0) -> Never { fatalError() }

func pass_2_noreturn() -> Int { evil_2() }
func pass_2_return() -> Int { return evil_2() }


print(pass_1_return())
print(pass_2_return())
print(pass_2_noreturn())
print(pass_1_noreturn())
```

Without implicit return:
pass_1_noreturn: traps
pass_1_return: returns 0
pass_2_noreturn: fails to compile (`error: missing return in a function expected to return 'Int'`)
pass_2_return: returns 0

With implicit return:
pass_1_noreturn: traps
pass_1_return: returns 0
pass_2_noreturn: returns 0
pass_2_return: returns 0

If I'm looking at this right, this seems to indicate that we get source compatibility without having to special case the scenario where the overload set includes a function which returns Never.  Is that right?

Reply:
That seems correct. I just don't like the two passes. Is it possible to say "if the expression type is uninhabited, write it back *and then skip checking the body?*" That way we're never type-checking the output of a type-check.

…And even with that, for performance reasons I don't like type-checking the same expression twice when there's not an error. Can we go the other way instead, and say "try to type-check this as if we're converting it to the return type, and if *that* fails try to type-check it as producing an uninhabited type, and if *that* fails print the diagnostics from the first system"?

Reply:
That makes a lot of sense, I'll try that out.

Reply:
Okay, as of 
https://github.com/apple/swift/pull/23251/commits/a1023f2aa475d4ad8d606ae3d7dca7d0d8dd8d86 , it solves the constraint system which the expression to be returned.  If that succeeds, the body is updated and the function returns.  Failing to solve that system, it tries to solve a second system where the expression is not returned.  If that succeeds, the expression is uninhabited and the expression is unequal to the function's return type, it deletes the return, updates the body, and returns.  If that fails, the diagnostics from attempting to salvage the first system are shown.  How does that compare with what you had in mind?

Reply:
Oof. That's exactly what I asked for and then I thought it through again and doesn't it break source compatibility in your `pass_1_noreturn` case above? That really is a case where there are multiple solutions and we need to pick the unusual one in order to not break source compatibility.

Reply:
Yeah, it does.  

We could just reorder the two cases, checking first whether we have an uninhabited solution to the system without a specified conversion type and then checking whether we have any solution to the system with a specified conversion type.  That unbreaks the `pass_1_noreturn` case.

It does, unfortunately, put us back into have two passes most of the time.  We can at least avoid type checking twice in the scenario where the expression is uninhabited, which gives the solution you mentioned above

> Is it possible to say "if the expression type is uninhabited, write it back and then skip checking the body?" That way we're never type-checking the output of a type-check.

How does that sound?

Reply:
:-( I don't think we can take a feature that requires type-checking something twice. That's still potentially a major slowdown if it's a really complex expression that *might* produce Never (but probably won't).

Reply:
I'll ask Doug and/or Pavel what they think tomorrow.

Reply:
That makes sense.  I'm going to look into the possibility of special casing overload resolution for single-expression functions.

Reply:
The specific special case I would think to try is "allow coercing T to the contextual type if T is uninhabited".

Reply:
A couple more data points:

"Allowing coercion from T to the contextual type if T is uninhabited" 

```
// ExprRewriter::coerceToType
  case TypeKind::Enum: {
    auto fromEnum = cast<EnumType>(desugaredFromType);
    if (fromEnum->isUninhabited()) {
      return cs.cacheType(new (tc.Context) UninhabitedUpcastExpr(expr, toType));
    }
    break;
```

requires that `solve` not be provided the function's result type as its `convertType`.  

```
// TypeChecker::typeCheckFunctionBodyUntil
      constraints::ConstraintSystem cs(*this, FD,
                                       constraints::ConstraintSystemFlags::SuppressDiagnostics);
      cs.setContextualType(E, TypeLoc::withoutLoc(resultType), CTP_ReturnStmt);

      SmallVector<constraints::Solution, 4> viable;
      if(!cs.solve(E, /*convertType*/Type(), /*listener=*/nullptr, viable,
                    FreeTypeVariableBinding::Disallow)) {
```

That works for those four fatalError overload cases above, but causes assertion failures when applying the solution on more basic things like

```
func wrong() -> String { 17 }
```

because the constraint system features an unhandled coercion:

```
Fixed score: 0 0 0 0 0 0 0 0 0 0 0 0
Type variables:
  $T0 as Int @ locator@0x7fb492039200 [IntegerLiteral@test2.swift:17:5]

Overload choices:

Constraint restrictions:

Disjunction choices:
Unhandled coercion
```

Taking that a step further, allowing uninhabited types to match arbitrary types

```
// ConstraintSystem::matchTypes
  if (kind >= ConstraintKind::Conversion) {
    if (type1->isUninhabited()) {
      conversionsOrFixes.push_back(ConversionRestrictionKind::UninhabitedUpcast);
    }
  }
```

(i.e. treating uninhabited types as bottom types, another language change) allows most of the new code in  `TypeChecker::typeCheckFunctionBodyUntil` to be deleted (and in particular type checking occurs once always), but continues to change the behavior of `pass_1_noreturn` if the solution is weighted:

```
ConstraintSystem::TypeMatchResult
ConstraintSystem::matchUninhabitedUpcastTypes(Type type1, Type type2,
                                              TypeMatchOptions flags,
                                              ConstraintLocatorBuilder locator) {
  if (type1->isUninhabited()) {
    increaseScore(SK_UninhabitedUpcast);
    return getTypeMatchSuccess();
  }

  assert(false);
  return getTypeMatchFailure(locator);
}
```

Dropping the weighting makes `pass_1_noreturn` behave as before but unacceptably changes the behavior of the solution of

```
func pass_1_return() -> Int { return evil_1() }
``` 

to resolve `evil_1` to the `Never` overload.

I'm looking at the feasibility of passing the fact that we're type-checking a single-expression function down into the constraint system so that we can not weight whenever return is omitted.

Reply:
@jrose-apple  Okay, I think I was able to resolve this: https://github.com/apple/swift/pull/23251/commits/a42c2c1a702432aa4c0add85b884b294c772ebc0 .  The change allows implicit conversion from an uninhabited type, but only from the type of a single expression of a function to the function's result type.  The conversion unlocks eliminating the less pleasant parts of the `TypeChecker::typeCheckFunctionBodyUntil`, leaving just the check for `Void` and the removal of the return in that case.  Just as importantly, it means that type-checking single expression functions goes through the same path as all other functions, giving us back all the diagnostics that get inserted along the way.